### DISCUSSION TITLE
getQueryData not updated on useQuery based on comp tree

### DISCUSSION THREAD

**Discussion by PaulKujawa** [TanStack/query#815] [General] [answered]
Example: https://codesandbox.io/s/react-query-cache-uhcpf?file=/src/App.tsx

If I request data in L32 the cache does not become updated after the server response arrived, whereas if I put `getQueryData` into L20 it works just fine. Is this a bug or do I miss something obvious? I am irritated because both places are "just" functional, parent components. I don't see how this should differ.

Please note that codesandbox might need you to reload the page to update the view. I encountered this issue in my project first. So it's not related to the online editor.

**Comment by wolverineks** [comment]
are you expecting `queryCache.getQueryData` to rerender the component with the data changes?

**Comment by PaulKujawa** [comment, marked as answer]
@wolverineks yes, that is what I expect(ed). Does this not work? How do my two cases differ? Any help would be appreciated.

**Reply by wolverineks** [reply]
`queryCache.getQueryData` is an imperative operation that does not subscribe to changes
feel free to use it when you want to get the most recent data, such as in a callback

for example, in optimistic updating, it is used to take a snapshot of the current data, update it immediately (before the promise with the actual value resolves), and rollback to the snapshot if the promise rejects

https://react-query.tanstack.com/docs/examples/optimistic-updates

**Reply by wolverineks** [reply]
now that i look closer, the issue might be caused by passing `children`
a component that is passed a `children` prop, will never rerender the children
`Comp2` does not rerender when `Comp1` rerenders, however, `Comp2` _does_ rerender when `App` rerenders
`Comp2` is not subscribed to the query, so it doesnt rerender when the query data changes
`Comp2` needs to rerender in order to display the new data

**Reply by wolverineks** [reply]
`Comp2` is not passed as `children` to `App`, so when `App` rerenders, so does `Comp2`

**Reply by PaulKujawa** [reply]
I am sorry for my shamefully late reply. You're absolutely right! Having `Comp1` render `Comp2` as a render prop would solve this issue as well. But then, if `queryCache.getQueryData` is only good for snapshot values, how does one render a response in multiple components on a page without requesting it for each component?

**Reply by TkDodo** [reply]
Every component that wants to subscribe to changes can just do `useGetPosts()`. This does not mean that multiple requests are sent to the backend.

**Reply by Delamern** [reply]
What if useGetPosts() requires additional parameters to run the query that Comp1 has but Comp2 does not? Like if useGetPosts signature was something like useGetPosts(arg1, arg2, arg3) where all params are used in the query, and Comp2 has no idea what any of those arguments are?

**Reply by TkDodo** [reply]
@Delamern I think even `Comp2` needs to be able to actually make the request to the backend - if it's the only component currently mounted, it will be making an actual request, so it needs to be aware of all the arguments to `useGetPosts`. Maybe a concrete example would help me understand this better. If I take a simple one - `usePost('posts', postId)` - every component that wants to display the posts details must be aware of it's id. Otherwise, how can it display it?

You can use any method of sharing client state (prop drilling, context, a library) to make the arguments available to all consumers.
